site/app/libraries/database/DatabaseQueries.php

```diff
@@ -2637,83 +2637,144 @@ public function getSubmitterIdFromAnonId(string $anon_id) {
+        $query = "SELECT user_id FROM users";
+        $this->course_db->query($query);
+        return $this->rowsToArray($this->course_db->rows());
     }
 
-    //TODO: refactor this to use push notification function..
-    public function pushSingleNotification($notification) {
-        $params = array();
-        $params[] = $notification->getComponent();
-        $params[] = $notification->getNotifyMetadata();
-        $params[] = $notification->getNotifyContent();
-        $params[] = $notification->getNotifySource();
-        $params[] = $notification->getNotifyTarget();
+    /**
+     * Get all users with a preference
+     * @param string $column
+     * @return array
+     */
+    public function getAllUsersWithPreference(string $column) {
+        $preferences = [
+            'merge_threads',
+            'all_new_threads',
+            'all_new_posts',
+            'all_modifications_forum',
+            'reply_in_post_thread',
+            'merge_threads_email',
+            'all_new_threads_email',
+            'all_new_posts_email',
+            'all_modifications_forum_email',
+            'reply_in_post_thread_email'
+        ];
+        $query = "SELECT user_id FROM notification_settings WHERE {$column} = 'true'";
+        $this->course_db->query($query);
+        if (!in_array($column,$preferences)) {
+            throw new DatabaseException("Given column, {$column}, is not a valid column", $query);
+        }
+        return $this->rowsToArray($this->course_db->rows());
+    }
+
+    /**
+     * Gets All Parent Authors who this user responded to
+     * @param string $post_author_id current_user_id
+     * @param string $post_id   the parent post id
+
+     */
+    public function getAllParentAuthors(string $post_author_id, string $post_id) {
+        $params = [$post_id, $post_author_id];
+        $query = "SELECT * FROM
+                  (WITH RECURSIVE parents AS (
+                  SELECT
+                    author_user_id, parent_id, id FROM  posts
+                  WHERE id = ?
+                  UNION SELECT
+                    p.author_user_id, p.parent_id, p.id
+                  FROM
+                    posts p
+                   INNER JOIN parents pa ON pa.parent_id = p.id
+                  ) SELECT DISTINCT 
+                    author_user_id AS user_id
+                  FROM
+                    parents
+                  WHERE author_user_id <> ?) AS parents;";
+        $this->course_db->query($query,$params);
+        return $this->rowsToArray($this->course_db->rows());
+    }
+
+    /**
+     * returns all authors who want to be notified if a post has been made in a thread they have posted in
+     * @param $thread_id
+     * @param $column ("reply_in_thread" or "reply_in_thread_email")
+     * @return array
+     */
+    public function getAllThreadAuthors($thread_id, $column) {
+        $params = [$thread_id];
+        $preferences = $this->getAllPreferences();
```

Comment:
is this defined anywhere?

Reply:
I could not find a relevant query.

Reply:
I'm not sure what that means? It looks like this line refers to a non-existent function in this file and will throw an error when used.